### PR TITLE
32# use database_url in prod

### PR DESCRIPTION
**Pull request by jmoldyvan**
Closes (#32)

# Description

In the files (.env .env.developement .env.sample .env.test) I added the DATABASE_URL component so Heroku can automatically update it and not break everything.
also made changes to the files server/lib/db.ts and server/config/config in envalid and initiating the migrations.

## Testing

make sure everything runs as normal in the developement. just remember to make the changes to personal .env in prod

## Type of change

changes to code

## Screenshots

Please include any screenshots if there are visual changes.

# Checklist:

- [ ] Changes have new/updated automated tests, if applicable
- [ ] Changes have new/updated docs, if applicable
- [X ] I have performed a self-review of my own code
- [X ] I have added comments on any new, hard-to-understand code
- [X ] Changes have been manually tested
- [ X] Changes generate no new errors or warnings

#Learned
I would like to mention i had attempted to use the envalid url validator as well as node.js URL to have an easier time parsing, however i ended up not needing to parse as Sequelize takes a URI parameter!

### CODE REVIEW

File: .env.sample [hunk 1]
```diff
@@ -1,4 +1,5 @@
 DISCORD_CLIENT_ID=1082239786162606080
 NEXT_PUBLIC_DISCORD_CLIENT_ID=1082239786162606080
 DISCORD_CLIENT_SECRET=123
-DB_PASSWORD=password
+DATABASE_URL= postgres://{user}:{password}@{hostname}:{port}/{database-name}
```
Comment (timmyichen) [inline .env.sample:4]: Are the curly braces here intentional?

File: server/lib/db.ts [hunk 2]
```diff
@@ -24,19 +25,29 @@ export class Database {
...
-      logging: env.DB_LOGGING,
-    });
+    if( env.isProd ){
+      this.sequelize = new Sequelize(env.DATABASE_URL,{
+        dialectOptions,
+        dialect: 'postgres',
+        logging: env.DB_LOGGING,
+      });
+    }
+    else{
```
Comment (timmyichen) [inline server/lib/db.ts:39]:
```suggestion
    if (env.isProd) {
      this.sequelize = new Sequelize(env.DATABASE_URL, {
        dialectOptions,
        dialect: 'postgres',
        logging: env.DB_LOGGING,
      });
    } else {
```

nit: formatting

i'll add lint rules later so this doesn't become something to address in PRs 🙃

File: server/lib/db.ts [hunk 3]
```diff
@@ -24,19 +25,29 @@ export class Database {
   dbName: string;
   sequelize: Sequelize;
 
-  constructor(database = env.DB_NAME) {
+  constructor(database= env.DB_NAME) {
```
Comment (timmyichen) [inline server/lib/db.ts:28]: same lint thing here - space before `=`

File: server/config/config.js [hunk 4]
```diff
@@ -4,11 +4,12 @@ const { cleanEnv, str } = require('envalid');
...
 
 const env = cleanEnv(process.env, {
-  DB_PORT: str(),
-  DB_USER: str(),
+  DATABASE_URL: str(),
   DB_HOST: str(),
   DB_NAME: str(),
+  DB_USER: str(),
   DB_PASSWORD: str(),
+  DB_PORT: str(),
```
Comment (timmyichen) [inline server/config/config.js:12]: why the reordering?

File: .env.test [hunk 5]
```diff
@@ -3,14 +3,17 @@ AUTH_SECRET=secret
...
 DB_NAME=test
 
+
 DISCORD_REDIRECT_URI=http://localhost:3000/auth-redirect
 DISCORD_GUILD_ID=735923219315425401
 DISCORD_CLIENT_ID=123
 
 NEXT_PUBLIC_DISCORD_CLIENT_ID=123
+
+DATABASE_URL= postgres://postgres:password@localhost:5432/test
```
Comment (timmyichen) [inline .env.test:19]: nit: newline

File: .env.test [hunk 6]
```diff
@@ -3,14 +3,17 @@ AUTH_SECRET=secret
 DISCORD_CLIENT_SECRET=123
 DB_PASSWORD=password
 
-# Configuration Values
+# Configuration Values# Configuration Values
```
Comment (timmyichen) [inline .env.test:6]: i think something got duplicated

File: .env.development [hunk 7]
```diff
@@ -1,8 +1,10 @@
...
+NEXT_PUBLIC_DISCORD_REDIRECT_URI=http://localhost:3000/auth-redirect
+DATABASE_URL= postgres://postgres:password@localhost:5432/dev-directory
 DB_PORT=5432
 DB_USER=postgres
 DB_HOST=localhost
 DB_NAME=dev-directory
-DISCORD_GUILD_ID=735923219315425401
-DISCORD_REDIRECT_URI=http://localhost:3000/auth-redirect
-NEXT_PUBLIC_DISCORD_REDIRECT_URI=http://localhost:3000/auth-redirect
+DB_PASSWORD=password
```
Comment (timmyichen) [inline .env.development:10]: why the reordering? also, nit: newline

File: .env.sample [hunk 8]
```diff
@@ -1,4 +1,5 @@
...
+DB_PASSWORD=password
```
Comment (timmyichen) [inline .env.sample:5]: nit: newline

File: server/lib/db.ts [hunk 9]
```diff
@@ -24,19 +25,29 @@ export class Database {
...
-      password: env.DB_PASSWORD,
-      host: env.DB_HOST,
-      database: database,
-      dialectOptions,
-      dialect: 'postgres',
-      logging: env.DB_LOGGING,
-    });
+    if( env.isProd ){
+      this.sequelize = new Sequelize(env.DATABASE_URL,{
+        dialectOptions,
```
Comment (timmyichen) [inline server/lib/db.ts:34]: this `dialectOptions` is also something that's prod-specific, so we can probably just include it here and then remove it from below

File: server/lib/db.ts [hunk 10]
```diff
@@ -24,19 +25,29 @@ export class Database {
...
+        dialect: 'postgres',
+        logging: env.DB_LOGGING,
```
Comment (timmyichen) [inline server/lib/db.ts:36]: And then these are common across both environments, so we can pull these out into a separate object and then spread it into the options here (to keep the code DRY)

**Review by timmyichen — CHANGES_REQUESTED** [review]

File: .env.sample (hunk 1 above)
Comment (devonzara) [reply .env.sample:4]: @timmyichen, thoughts on simplifying the code and just using `DATABASE_URL` in all environments? We can throw it at the bottom of the `.env` and automatically build it using the separate vars. Heroku can then overwrite `DATABASE_URL` all they want and the separate vars would simply no longer be used/referenced anywhere. eg.

```ini
...
DB_PASSWORD=password

# Configuration Values# Configuration Values
DB_PORT=5432
DB_USER=postgres
DB_HOST=localhost
DB_NAME=test

...

DATABASE_URL= postgres://$DB_USER:$DB_PASSWORD@$DB_HOST:$DB_PORT/$DB_NAME
```

[Var Expansion in NextJS Docs](https://nextjs.org/docs/basic-features/environment-variables#:~:text=automatically%20expand%20variables%20(%24VAR))

Comment (jmoldyvan) [reply .env.sample:4]: yeah, it was, but I can change it?

Comment (timmyichen) [reply .env.sample:4]: I guess as a whole we can probably remove it since it'll never be used in development

**Comment by jmoldyvan** [comment]
fixes all the formatting!! the recording was me getting ahead of myself when posting the pr

**Comment by timmyichen** [comment]
> fixes all the formatting!! the recording was me getting ahead of myself when posting the pr

Thanks!  Just a few outstanding comments/questions, plus if you look at the changed files and you see something like this, it means you need the file to end in a blank newline :) 
image

**Review by MainlyColors — APPROVED** [review]
LGTM

File: .env.sample (hunk 1 above)
Comment (jmoldyvan) [reply .env.sample]: gotcha!

**Comment by jmoldyvan** [comment]
alright I think I fixed all the comments!

File: .env.development [hunk 11]
```diff
@@ -6,3 +6,4 @@ DB_NAME=dev-directory
 DISCORD_GUILD_ID=735923219315425401
 DISCORD_REDIRECT_URI=http://localhost:3000/auth-redirect
 NEXT_PUBLIC_DISCORD_REDIRECT_URI=http://localhost:3000/auth-redirect
+DATABASE_URL= postgres://postgres:password@localhost:5432/dev-directory
```
Comment (timmyichen) [inline .env.development:9]: one last thing 🙃 there shouldn't be a space after `=` in env files

File: .env.test [hunk 12]
```diff
@@ -14,3 +14,5 @@ DISCORD_GUILD_ID=735923219315425401
 DISCORD_CLIENT_ID=123
 
 NEXT_PUBLIC_DISCORD_CLIENT_ID=123
+
+DATABASE_URL= postgres://postgres:password@localhost:5432/test
```
Comment (timmyichen) [inline .env.test:18]: same thing here

File: .env.development [hunk 13]
```diff
@@ -6,3 +6,4 @@ DB_NAME=dev-directory
 DISCORD_GUILD_ID=735923219315425401
 DISCORD_REDIRECT_URI=http://localhost:3000/auth-redirect
 NEXT_PUBLIC_DISCORD_REDIRECT_URI=http://localhost:3000/auth-redirect
+DATABASE_URL = postgres://postgres:password@localhost:5432/dev-directory
```
Comment (timmyichen) [inline .env.development:9]: oop - i mean, in this file we want there to be _no_ space (see examples on previous line) 🙃

**Review by timmyichen — APPROVED** [review]
🎉 thanks for dealing with the constant requests :)

**Pull request merged**